pennylane/fourier/reconstruct.py

```diff
@@ -0,0 +1,523 @@
+    of frequencies, ``nums_frequency`` , per parameter must be provided.
+
+    **Usage Details**
+
+    *Input formatting*
+
+    As described briefly above, the essential inputs to ``reconstruct`` that provide information
+    about the QNode are given as dictionaries of dictionaries, where the outer keys reference
+    the argument names of ``qnode`` and the inner keys reference the parameter indices within
+    one array-valued QNode argument. For scalar-valued QNode parameters, the parameter index is
+    set to ``0`` by convention (also see the examples below).
+    This applies to ``nums_frequency`` , ``spectra`` , and ``shifts`` .
+
+    Note that the information provided in ``nums_frequency`` / ``spectra`` is essential for
+    the correctness of the reconstruction.
+
+    On the other hand, the input format for ``ids`` is flexible and allows a collection of
+    parameter indices for each QNode argument name (as a ``dict`` ), a collection of argument
+    names (as a ``list``, ``set``, ``tuple`` or similar), or a single argument name
+    (as a ``str`` ) to be defined. For ``ids=None`` , all argument names contained in
+    ``nums_frequency`` -- or ``spectra`` if ``nums_frequency`` is not used -- are considered.
+    For inputs that do not specify parameter indices per QNode argument name (all formats but
+    ``dict`` ), these parameter indices are inferred from ``nums_frequency`` / ``spectra`` .
+
+    *Reconstruction cost*
+
+    The reconstruction cost -- in terms of calls to ``qnode`` -- depend on the number of
+    frequencies given via ``nums_frequency`` or ``spectra`` . A univariate reconstruction
+    for :math:`R` frequencies takes :math:`2R+1` evaluations. If multiple univariate
+    reconstructions are performed at the same point with various numbers of frequencies
+    :math:`R_k` , the cost are :math:`1+2\sum_k R_k` if the shift :math:`0` is used in all
+    of them. This is in particular the case if ``nums_frequency`` or ``spectra`` with
+    ``shifts=None`` is used.
+
+    If the number of frequencies is too large or the given frequency spectrum contains
+    more than the spectrum of ``qnode`` , the reconstruction is performed suboptimally
+    but remains correct.
+    For integer-valued spectra with gaps, the equidistant reconstruction is thus suboptimal
+    and the non-equidistant version method be used.
+
+    *Numerical stability*
+
+    In general, the reconstruction with equidistant shifts for equidistant frequencies
+    (used if ``nums_frequency`` is provided) is more stable numerically than the more
+    general Fourier reconstruction (used if ``nums_frequency=None`` ).
+    If the system of equations to be solved in the Fourier transform is
+    ill-conditioned, a warning is raised as the output might become unstable.
+    Examples for this are shift values or frequencies that lie very close to each other.
+
+    **Examples**
+
+    Consider the following QNode:
```

Comment:
For the 'minimal' QNode example, I'm wondering if it should be intrinsically linked to `qnode_spectrum`? E.g.,

```python
@qml.qnode(dev)
def circuit(x, Y, f=1.0):
    qml.RX(f * x, wires=0)
    qml.RY(Y[0], wires=0)
    qml.RY(Y[1], wires=1)
    qml.CNOT(wires=[0, 1])
    qml.RY(5 * Y[1], wires=1)
    return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))

x = np.array(0.4, requires_grad=True)
Y = np.array([1.9, -0.5], requires_grad=True)
f = np.array(2.3, requires_grad=True)
```

Generate reconstruction:

```pycon
>>> spectrum = qml.fourier.qnode_spectrum(circuit)(x, Y)
>>> recon_fns = qml.fourier.reconstruct(circuit, spectra=spectrum)(x, Y)
>>> recon_fns
{'x': {(): <function _reconstruct_gen.<locals>._reconstruction at 0x7f7f099334c0>}, 'Y': {(0,): <function _reconstruct_gen.<locals>._reconstruction at 0x7f7f09988e50>, (1,): <function _reconstruct_gen.<locals>._reconstruction at 0x7f7f09933c10>}}
```

Evaluation:

```pycon
>>> recon_x = recon_fns["x"][0]
>>> recon_x(x + 0.1)
-0.6216658475465537
>>> circuit(x + 0.1, Y)
-0.6216658475465537
```

Gradients:

```pycon
>>> qml.grad(circuit, argnum=0)(x + 0.1, Y)
-0.04447110901737461
>>> qml.grad(recon_x)(x + 0.1)
-0.04447110901737466
```

Reply:
You could also show that calling `recon_x` causes no circuit executions 😍

Reply:
The additional examples below, since being slightly more complex, could be moved into `.. UsageDetails::`?

Reply:
I included a link to `qnode_spectrum` in the advanced `.. UsageDetails::` to keep the first example truly minimal :)